virtcontainers/network.go

```diff
@@ -683,41 +683,67 @@ func bridgeNetworkPair(endpoint Endpoint, numCPUs uint32, disableVhostNet bool)
 		return fmt.Errorf("Could not set TAP MTU %d: %s", attrs.MTU, err)
 	}
 
-	hardAddr, err := net.ParseMAC(netPair.VirtIface.HardAddr)
-	if err != nil {
+	if err := netHandle.LinkSetUp(tapLink); err != nil {
+		return fmt.Errorf("Could not enable TAP %s: %s", netPair.TAPIface.Name, err)
+	}
+
+	tapAttrs := tapLink.Attrs()
+
+	if err := addQdiscIngress(tapAttrs.Index); err != nil {
 		return err
 	}
-	if err := netHandle.LinkSetHardwareAddr(link, hardAddr); err != nil {
-		return fmt.Errorf("Could not set MAC address %s for veth interface %s: %s",
-			netPair.VirtIface.HardAddr, netPair.VirtIface.Name, err)
+
+	if err := addQdiscIngress(attrs.Index); err != nil {
+		return err
 	}
 
-	mcastSnoop := false
-	bridgeLink, _, err := createLink(netHandle, netPair.Name, &netlink.Bridge{MulticastSnooping: &mcastSnoop}, int(numCPUs))
-	if err != nil {
-		return fmt.Errorf("Could not create bridge: %s", err)
+	if err := addRedirectTCFilter(attrs.Index, tapAttrs.Index); err != nil {
+		return err
 	}
 
-	if err := netHandle.LinkSetMaster(tapLink, bridgeLink.(*netlink.Bridge)); err != nil {
-		return fmt.Errorf("Could not attach TAP %s to the bridge %s: %s",
-			netPair.TAPIface.Name, netPair.Name, err)
+	if err := addRedirectTCFilter(tapAttrs.Index, attrs.Index); err != nil {
+		return err
 	}
 
-	if err := netHandle.LinkSetUp(tapLink); err != nil {
-		return fmt.Errorf("Could not enable TAP %s: %s", netPair.TAPIface.Name, err)
+	return nil
+}
+
+func addQdiscIngress(index int) error {
```

Comment:
What happens if `index < 0`? Does `netlink.QdiscAdd()` handle that? Is it a valid index? (might be given `QdiscAttrs.Index` is an `int` wheres all other members are `uint32`?!?)